### PR TITLE
nifi: init at 1.16.0

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Add Apache NiFI version 1.16.0.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/nifi/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{ lib, stdenv, fetchurl, makeWrapper, jre8 }:
+
+stdenv.mkDerivation rec {
+  pname = "nifi";
+  version = "1.14.0";
+
+  src = fetchurl {
+    url = "https://dlcdn.apache.org/nifi/1.14.0/nifi-1.14.0-bin.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/servers/nifi/default.nix:8]: Please substitute the version numbers.

File: pkgs/servers/nifi/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      --replace "/bin/sh" "${stdenv.shell}"
+    substituteInPlace $out/bin/nifi-env.sh \
+      --replace "#export JAVA_HOME=/usr/java/jdk1.8.0/" "export JAVA_HOME=${jre8}"
+  '';
+
+  meta = with lib; {
+    description = ''
+      Apache NiFi supports powerful and scalable directed graphs of data routing,
+      transformation, and system mediation logic.
+    '';
```
Comment (SuperSandro2000) [inline pkgs/servers/nifi/default.nix:34]: That is to long for description. Maybe move it to longDescription?

**Comment by Izorkin** [comment]
@SuperSandro2000 thanks, fixed.

**Comment by Izorkin** [comment]
сс @aanderse @flokli @Mic92

File: pkgs/servers/nifi/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    mv $out/{LICENSE,NOTICE,README} $out/share/nifi
+
+    substituteInPlace $out/bin/nifi.sh \
+      --replace "/bin/sh" "${stdenv.shell}"
+    substituteInPlace $out/bin/nifi-env.sh \
+      --replace "#export JAVA_HOME=/usr/java/jdk1.8.0/" "export JAVA_HOME=${jre8}"
+  '';
+
+  meta = with lib; {
+    longDescription = ''
```
Comment (SuperSandro2000) [inline pkgs/servers/nifi/default.nix:31]: Missing descripiton.

File: pkgs/servers/nifi/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{ lib, stdenv, fetchurl, makeWrapper, jre8 }:
+
+stdenv.mkDerivation rec {
+  pname = "nifi";
+  version = "1.14.0";
+
+  src = fetchurl {
+    url = "https://dlcdn.apache.org/${pname}/${version}/${pname}-${version}-bin.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/servers/nifi/default.nix:8]:
```suggestion
    url = "https://dlcdn.apache.org/nifi/${version}/nifi-${version}-bin.tar.gz";
```

**Comment by Izorkin** [comment]
Fixed PR.

**Comment by Izorkin** [comment]
@GrahamcOfBorg eval

File: nixos/modules/services/web-apps/nifi.nix [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+          test -f '/var/lib/nifi/conf/state-management.xml'                 || (cp '${cfg.package}/share/nifi/conf/state-management.xml' '/var/lib/nifi/conf' && chmod 0640 '/var/lib/nifi/conf/state-management.xml')
+          test -f '/var/lib/nifi/conf/zookeeper.properties'                 || (cp '${cfg.package}/share/nifi/conf/zookeeper.properties' '/var/lib/nifi/conf' && chmod 0640 '/var/lib/nifi/conf/zookeeper.properties')
+          ${optionalString ((cfg.initUser!=null) && (cfg.initPasswordFile!=null)) ''
+            awk -F'[<|>]' '/property name="Username"/ {if ($3!="") f=1} END{exit !f}' /var/lib/nifi/conf/login-identity-providers.xml || ${cfg.package}/bin/nifi.sh set-single-user-credentials ${cfg.initUser} $(cat ${cfg.initPasswordFile})
+          ''}
+          ${optionalString (cfg.enableHTTPS == false) ''
+            awk -F'=' '/^nifi.remote.input.secure=/ {if ($2=="false") f=1} END{exit !f}' /var/lib/nifi/conf/nifi.properties                   || sed -i -e 's|nifi.remote.input.secure=.*|nifi.remote.input.secure=false|g' /var/lib/nifi/conf/nifi.properties
```
Comment (Mic92) [inline nixos/modules/services/web-apps/nifi.nix:151]: This looks super hacky. Is there no better way to write these properties with less code?

File: nixos/modules/services/web-apps/nifi.nix [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      environment = env;
+      path = [ pkgs.gawk ];
+
+      serviceConfig = {
+        Type = "forking";
+        PIDFile = "/run/nifi/nifi.pid";
+        ExecStartPre = pkgs.writeScript "nifi-pre-start.sh" ''
+          #!/bin/sh
+          umask 077
+          test -f '/var/lib/nifi/conf/authorizers.xml'                      || (cp '${cfg.package}/share/nifi/conf/authorizers.xml' '/var/lib/nifi/conf' && chmod 0640 '/var/lib/nifi/conf/authorizers.xml')
```
Comment (Mic92) [inline nixos/modules/services/web-apps/nifi.nix:136]: There is also a lot of repetitive code here.

**Review by Mic92 — CHANGES_REQUESTED** [review]
I think the preStart code should use some function so it does not repeat so much code.

File: nixos/modules/services/web-apps/nifi.nix (hunk 5 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/nifi.nix:151]: This is a simple option that I managed to implement.
If generate a complete configuration file, then will not be able to manually change the parameters that are not specified in this module.
In this variant, we can change the main parameters, and the rest of the parameters can be changed manually.

Need a way that takes as a basis the default configuration file and changes certain parameters in it.

Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/nifi.nix:151]: There needs to be a better option than this. At least I think we should move this to an extra file.

Comment (Izorkin) [reply nixos/modules/services/web-apps/nifi.nix:151]: Allowed this variant?:
```
...
          ${optionalString ((cfg.initUser!=null) && (cfg.initPasswordFile!=null)) ''
            ${cfg.package}/bin/nifi.sh set-single-user-credentials ${cfg.initUser} $(cat ${cfg.initPasswordFile})
          ''}
          ${optionalString (cfg.enableHTTPS == false) ''
            sed -i -e 's|nifi.remote.input.secure=.*|nifi.remote.input.secure=false|g' /var/lib/nifi/conf/nifi.properties
            sed -i -e 's|nifi.web.http.host=.*|nifi.web.http.host=${cfg.listenHost}|g' /var/lib/nifi/conf/nifi.properties
            sed -i -e 's|nifi.web.http.port=.*|nifi.web.http.port=${(toString cfg.listenPort)}|g' /var/lib/nifi/conf/nifi.properties
...
```

File: pkgs/servers/nifi/default.nix [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    substituteInPlace $out/bin/nifi.sh \
+      --replace "/bin/sh" "${stdenv.shell}"
+    substituteInPlace $out/bin/nifi-env.sh \
+      --replace "#export JAVA_HOME=/usr/java/jdk1.8.0/" "export JAVA_HOME=${jre8}"
+  '';
+
+  meta = with lib; {
+    description = ''
+      An easy to use, powerful, and reliable system to process and distribute data.
+    '';
```
Comment (SuperSandro2000) [inline pkgs/servers/nifi/default.nix:33]:
```suggestion
    description = "Easy to use, powerful, and reliable system to process and distribute data";
```

File: nixos/modules/services/web-apps/nifi.nix (hunk 5 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/nifi.nix]: Updated `preStart` script.

File: pkgs/servers/nifi/default.nix (hunk 7 above)
Comment (Izorkin) [reply pkgs/servers/nifi/default.nix]: Fixed

**Comment by Izorkin** [comment]
cc @SuperSandro2000 @Mic92

**Comment by Lassulus** [comment]
for a new module a test would be nice?

**Comment by Izorkin** [comment]
I don't know how to check working service through cli.

**Comment by Lassulus** [comment]
hmm, maybe a minimal test which just checks if service starts?

**Comment by Izorkin** [comment]
Ok, i'll try to do a test.

**Comment by Izorkin** [comment]
@Lassulus creating small test. The problem is that NiFi takes a very long time to start up. Test fails due to timeout.

**Comment by Lassulus** [comment]
> @Lassulus creating small test. The problem is that NiFi takes a very long time to start up. Test fails due to timeout.

hmm well than maybe no test

**Comment by aanderse** [comment]
How long? I thought we could tweak the timeout if needed.

**Comment by Izorkin** [comment]
> How long? I thought we could tweak the timeout if needed.

Initial startup takes more than 10-12 minutes. Added  option `meta.timeout = 18000` - test terminates with error:
```
Exception: action timed out after 900 seconds
```

**Comment by aanderse** [comment]
oof ...

> hmm well than maybe no test

👍

Thanks for checking into it @Izorkin!

**Comment by Izorkin** [comment]
I will update PR soon.

**Comment by Izorkin** [comment]
I will update PR soon.

**Comment by Izorkin** [comment]
@aanderse @Lassulus updated PR!

**Review by Lassulus — APPROVED** [review]

**Comment by Izorkin** [comment]
Created test, I will add soon.

**Comment by Izorkin** [comment]
Added nifi test. Should have increased the size of disk.

**Comment by Lassulus** [comment]
nice work

**Pull request merged**